### DISCUSSION TITLE
GPS Sensor

### DISCUSSION THREAD

**Discussion by sanjay-swain** [google-deepmind/mujoco#1712] [Ideas]
Are there any plans for implementing a GPS (Global Positioning Sensor) sensor in mujoco? It would be great to be able to simulate GPS coordinates (latitude, longitude) in the virtual world.

With GPS I am pretty sure flight control software like [ardupilot](https://ardupilot.org/) can be used with the simulator as Mujoco already have magnetometer, and imu sensors.

**Comment by yuvaltassa** [comment]
How would that work though? GPS is based on the curved Earth frame, while the MuJoCo world frame is standard Cartesian...

**Reply by sanjay-swain** [reply]
Yeah. We need to convert Cartesian coordinates to WGS84. Gazebo has a `NavSat` sensor which does this and webots also have GPS sensors.

These sensors can be used to simulate some autonomous vehicles like drones and rovers with waypoint navigation through GPS coordinates.

In Gazebo you can set what your starting coordinates will be and it will emulate the data received from a GPS sensor.

I will update this reply when I have some time for some resources on how it is implemented in Gazebo and Webots.

So I think this sensor could be useful for some testing of robots in larger scale and it should also be easy to just create a simple sensor plugin isn't it?

**Reply by yuvaltassa** [reply]
A sensor plugin would indeed be easy.

I'm curious to understand how Gazebo and Webots do this. Do they really a have a sphere the size of the Earth? Variable gravity?

**Reply by sanjay-swain** [reply]
Oh no they don't simulate entire earth. They also have flat plane exactly like mujoco. The GPS sensor just converts the cartisian coordinates from the simulation to WGS84, which is latitude and longitude. WGS84 is a type of coordinate system where location is defined on a 3 dimensional sphere.

Basically they take the normal plane and mathematically "wraps" it around a sphere to calculate what would be the latitude and longitude given the starting latitude and longitude.

To do this in webots or gazebo first we setup what would be the global coordinates (lat and lon) of the origin of simulation and from there we can calculate the global cordinate of any position in 2D plane.

Check out this [matlab function](https://www.mathworks.com/help/map/ref/ned2geodetic.html). NED means North East Down, this convention is used in aviation where INS (Inertial Navigation System) calculate the current position. And geodetic is basically the latitude and longitude.

**Reply by yuvaltassa** [reply]
Can you paste the code here?
